google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +386,48 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        exactlyOncePendingBatch.add(outstandingMessage);
+        outstandingReceipts.put(message.getAckId(), outstandingMessage);
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if(outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      OutstandingMessage outstandingMessage = outstandingReceipts.get(ackRequestData.getAckId());
+      pendingMessages.putIfAbsent(outstandingMessage.receivedMessage.getAckId(),
```

Comment:
If we can't insert into pending messages here, we know a message is a duplicate, so we can continue